java/source/infodynamics/measures/continuous/kraskov/MultiVariateInfoMeasureCalculatorKraskov.java

```diff
@@ -0,0 +1,538 @@
+    lastAverage = computeFromObservations(false)[0];
+    isComputed = true;
+    if (debug) {
+      Calendar rightNow2 = Calendar.getInstance();
+      long endTime = rightNow2.getTimeInMillis();
+      System.out.println("Calculation time: " + ((endTime - startTime)/1000.0) + " sec" );
+    }
+    return lastAverage;
+  }
+
+  /**
+   * {@inheritDoc}
+   * 
+   * @return the "time-series" of local measure values in nats (not bits!)
+   * @throws Exception
+   */
+  public double[] computeLocalOfPreviousObservations() throws Exception {
+    double[] localValues = computeFromObservations(true);
+    lastAverage = MatrixUtils.mean(localValues);
+    isComputed = true;
+    return localValues;
+  }
+
+  /**
+   * This method, specified in {@link MultiVariateInfoMeasureCalculatorCommon}
+   * is not implemented yet here.
+   */
+  public double[] computeLocalUsingPreviousObservations(double[][] states) throws Exception {
+    // TODO If this is implemented, will need to normalise the incoming
+    //  observations the same way that previously supplied ones were
+    //  normalised (if they were normalised, that is)
+    throw new Exception("Local method not implemented yet");
+  }
+
+  /**
+   * This protected method handles the multiple threads which
+   *  computes either the average or local measure (over parts of the total
+   *  observations), computing the
+   *  distances between all tuples in time.
+   * 
+   * <p>The method returns:<ol>
+   *  <li>for (returnLocals == false), an array of size 1,
+   *      containing the average measure </li>
+   *  <li>for (returnLocals == true), the array of local
+   *      measure values</li>
+   *  </ol>
+   * 
+   * @param returnLocals whether to return an array or local values, or else
+   *  sums of these values
+   * @return either the average measure, or array of local measure value,
+   *  in nats not bits
+   * @throws Exception
+   */
+  protected double[] computeFromObservations(boolean returnLocals) throws Exception {
+    
+    double[] returnValues = null;
+
+    ensureKdTreesConstructed();
+    
+    if (numThreads == 1) {
+      // Single-threaded implementation:
+      returnValues = partialComputeFromObservations(0, totalObservations, returnLocals);
+      
+    } else {
+      // We're going multithreaded:
+      if (returnLocals) {
+        // We're computing locals
+        returnValues = new double[totalObservations];
+      } else {
+        // We're computing average
+        returnValues = new double[1];
+      }
+      
+      // Distribute the observations to the threads for the parallel processing
+      int lTimesteps = totalObservations / numThreads; // each thread gets the same amount of data
+      int res = totalObservations % numThreads; // the first thread gets the residual data
+      if (debug) {
+        System.out.printf("Computing Kraskov Multi-Info with %d threads (%d timesteps each, plus %d residual)\n",

```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.